docs/source-fabric/advanced/model_parallel/tp.rst

```diff
@@ -0,0 +1,279 @@
+        import torch.nn as nn
+        import torch.nn.functional as F
+
+        from torch.distributed.tensor.parallel import ColwiseParallel, RowwiseParallel
+        from torch.distributed.tensor.parallel import parallelize_module
+
+        import lightning as L
+        from lightning.pytorch.demos.boring_classes import RandomDataset
+        from lightning.fabric.strategies import ModelParallelStrategy
+
+
+        class FeedForward(nn.Module):
+            def __init__(self, dim, hidden_dim):
+                super().__init__()
+                self.w1 = nn.Linear(dim, hidden_dim, bias=False)
+                self.w2 = nn.Linear(hidden_dim, dim, bias=False)
+                self.w3 = nn.Linear(dim, hidden_dim, bias=False)
+
+            def forward(self, x):
+                return self.w2(F.silu(self.w1(x)) * self.w3(x))
+
+
+        def parallelize_feedforward(model, device_mesh):
+            # Lightning will set up a device mesh for you
+            tp_mesh = device_mesh["tensor_parallel"]
+            # Use PyTorch's distributed tensor APIs to parallelize the model
+            plan = {
+                "w1": ColwiseParallel(),
+                "w2": RowwiseParallel(),
+                "w3": ColwiseParallel(),
+            }
+            parallelize_module(model, tp_mesh, plan)
+            return model
+
+
+        strategy = ModelParallelStrategy(parallelize_fn=parallelize_feedforward)
+        fabric = L.Fabric(accelerator="cuda", devices=2, strategy=strategy)
+        fabric.launch()
+
+        # Initialize the model
+        model = FeedForward(8192, 8192)
+        model = fabric.setup(model)
+
+        # Define the optimizer
+        optimizer = torch.optim.AdamW(model.parameters(), lr=3e-3, foreach=True)
+        optimizer = fabric.setup_optimizers(optimizer)
+
+        # Define dataset/dataloader
+        dataset = RandomDataset(8192, 64)
+        dataloader = torch.utils.data.DataLoader(dataset, batch_size=8)
+        dataloader = fabric.setup_dataloaders(dataloader)
+
+        # Simplified training loop
+        for i, batch in enumerate(dataloader):
+            output = model(batch)
+            loss = output.sum()
+            fabric.backward(loss)
+            optimizer.step()
+            optimizer.zero_grad()
+            fabric.print(f"Iteration {i} complete")
+
+        fabric.print(f"Peak memory usage: {torch.cuda.max_memory_allocated() / 1e9:.02f} GB")
+
+|
+
+When measuring the peak memory consumption, we should see that doubling the number of GPUs reduces the memory consuption roughly by half:
+
+
+.. list-table::
+   :widths: 20 20 20 20 20
+   :header-rows: 1
+
+   * -
+     - 1 GPU (no TP)
+     - 2 GPUs
+     - 4 GPUs
+     - 8 GPUs
+   * - Memory per GPU
+     - 4.04 GB
+     - 2.03 GB
+     - 1.02 GB
+     - 0.60 GB
+
+Beyond this toy example, we recommend you study our `LLM Tensor Parallel Example (Llama 2) <https://github.com/Lightning-AI/pytorch-lightning/tree/master/examples/fabric/tensor_parallel>`_.
```

Comment:
This is both Llama 2 and Llama 3 right from TorchTitan right? If they are the same we could just say Llama 3 to make it more up to date.

Reply:
That's right, they are architecturally the same so we can name this Llama 3 👍